test/test_unary_ufuncs.py

```diff
@@ -378,6 +397,61 @@ def test_batch_vs_slicing(self, device, dtype, op):
 
         self.assertEqual(actual, expected)
 
+    def _compare_out_variant_based_on_device(self, op, input, out):
+        out_dtype = out.dtype
+        if self.device_type == 'cuda':
+            # As input are dynamically casted
+            # in kernel, cast input to out_dtype
+            expected = op(input.to(out_dtype))
+            op(input, out=out)
+            self.assertEqual(out, expected)
+        else:
```

Comment:
There should be no observable different based on the device type.

Reply:
I think the difference is because when it is CUDA, the input is dynamically cast to output type correctly.
So if input is `int64` and output is `double`, then `int64` is converted to `double` and output is saved in `double` tensor.

However on CPU, `int64` is converted to `float` as that is the computed common_dtype when integer is promoted and then we explicitly call `iter.cast_outputs()` (as discussed as per the offline conversation) to cast it to `double`.

Reference : https://github.com/pytorch/pytorch/blob/a277c097acc615b4710d615387891b7f764df326/aten/src/ATen/native/TensorIterator.cpp#L345-L347

I was facing tolerance issue on tests otherwise.

Let me know what you think.

Reply:
The common dtype is computed without considering the output, however, and the computation should occur in the common dtype. @zasdfgbnm, is there a bug in our CUDA tensor iterator loops where they're computing in the output dtype?

Reply:
Let's look at some example programs.

```
a = 7.3333484203230823
b = 2.44040844

print(a/b)

# Creates a double scalar tensor representable in half precision
a_t = torch.tensor(a, dtype=torch.double).half().double()
b_t = torch.tensor(b, dtype=torch.double).half().double()

print(a_t / b_t)

out = torch.empty(0, dtype=torch.double)
torch.div(a_t.half(), b_t.half(), out=out)
print(out)
```

So this program divides two values in Python, double, and then in half but with a double output. It prints:

```
3.0049676521865667
tensor(3.0056, dtype=torch.float64)
tensor(3.0059, dtype=torch.float64)
```

This is expected. Note the divergence between the double and half inputs. Also note that this divergence is not a representational issue with inputs because of the half->double bobble performed earlier.

Now let's do this on CUDA:

```
a = 7.3333484203230823
b = 2.44040844

print(a/b)

a_t = torch.tensor(a, dtype=torch.double).half().double().cuda()
b_t = torch.tensor(b, dtype=torch.double).half().double().cuda()

print(a_t / b_t)

out = torch.empty(0, dtype=torch.double).cuda()
torch.div(a_t.half(), b_t.half(), out=out)
print(out)
```

The program is the same except the tensors are placed on a CUDA device. This prints:

```
3.0049676521865667
tensor(3.0056, device='cuda:0', dtype=torch.float64)
tensor(3.0059, device='cuda:0', dtype=torch.float64)
```

The same values. So it appears that the output dtype is correctly not affecting the computation on CPU or CUDA for div.

Reply:
IIUC TensorIterator GPU loops allow doing computing in any dtype. The dynamic casting mechanism will handle the cast automatically. You can even do
float --(cast)--> int --(compute)--> int --(cast)--> float, 
nothing will prevent you from doing so, and there will be no error message. It is the responsibility of the operator implementor to ensure the correct type is used.
From https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/UnaryGeometricKernels.cu, we can see that lots of computations are done in output type, instead of common type.

Reply:
If we agree that computation should be done in common type instead of output type, then this file should be fixed.

Reply:
Agreed. I've only fixed the behavior in TensorIterator, and I'm sure other parts of our system still use the output argument to determine what computation they do. Analogously, our proposed updates to out resizing behavior were implemented in TensorIterator but few other places. Would you file an issue where you notice the out type affecting the computation, @zasdfgbnm?

Reply:
https://github.com/pytorch/pytorch/issues/46302

Reply:
> #46302

Thank you!

Reply:
OK, that conversation aside, what is going on in this particular case that the CPU and CUDA code paths are divergent, @kshitij12345? Is there an error with setting the appropriate computation dtype? Or a numerical precision issue on CUDA?

Reply:
It is the case of numerical difference.

Reply:
Oh, in that case would you just set the precision high enough for both of them?

Reply:
Is it fine if we update the CUDA computation type to common_dtype type as mentioned in the conversation above. 
I have verified locally that it solves the issue.
```C++
void sin_kernel_cuda(TensorIterator& iter) {
  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.common_dtype(), "sin_cuda", [&]() {
    gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
      return ::sin(a);
    });
  });
}
```

Reply:
@mruberry Do let me know if this is fine.

Reply:
Yes! I was just about to come here and say exactly that. We should fix this as @zasdfgbnm suggests above.

Reply:
Great. Done!